### PR TITLE
Enabled Navigation Prompting in Disclosed Content

### PR DESCRIPTION
**Pull request by tbiethman**
### Summary
We updated to update the terra-disclosure-manager to support pending action workflows. We also wanted to add support for the loading overlays/error boundaries provided by terra-application. In order to make those enhancements passively, we had to move the logic for the Disclosure-/Modal-/SlidePanelManager into the terra-application repository.

The passivity concern was around the addition of new peerDependencies. The APIs themselves are passive with the existing packages.

I also updated the dev-site configuration to match what we're doing in some other packages and ported some docs over.

### CODE REVIEW

**Comment by tbiethman** [comment]
Adding 'do not merge' label until terra-framework has been released.

File: tests/wdio/slide-panel-manager/slide-panel-manager-spec.js
Comment (tbiethman) [inline tests/wdio/slide-panel-manager/slide-panel-manager-spec.js]: New tests here

File: tests/wdio/modal-manager/modal-manager-spec.js
Comment (tbiethman) [inline tests/wdio/modal-manager/modal-manager-spec.js]: New tests here

**Review by StephenEsser — APPROVED** [review]

File: src/modal-manager/ModalManager.module.scss
Comment (ryanthemanuel) [inline src/modal-manager/ModalManager.module.scss]: Should these be --terra-application-modal-manager-border etc?

**Review by ryanthemanuel — APPROVED** [review]
Approved once the themeable stuff is figured out

**Review by mjhenkes — APPROVED** [review]

File: tests/wdio/slide-panel-manager/slide-panel-manager-spec.js
Comment (emilyrohrbough) [inline tests/wdio/slide-panel-manager/slide-panel-manager-spec.js]: The disclose width screenshots don't appear to be different. JK thats for the tiny viewport

File: src/terra-dev-site/reference/components/example/ApplicationLoadingOverlayProps.jsx
Comment (emilyrohrbough) [inline src/terra-dev-site/reference/components/example/ApplicationLoadingOverlayProps.jsx]: Why is this being rendered as a component?

File: src/disclosure-manager/DisclosureContainer.jsx
Comment (emilyrohrbough) [inline src/disclosure-manager/DisclosureContainer.jsx]: I don't think the `promptCheckpointRef` is needed. The NavigationPromptCheckpoint manages the list of registered props and will resolve from resolvePrompts:
https://github.com/cerner/terra-framework/blob/master/packages/terra-navigation-prompt/src/NavigationPromptCheckpoint.jsx#L121

File: src/modal-manager/ModalManager.jsx
Comment (emilyrohrbough) [inline src/modal-manager/ModalManager.jsx]: I don't think this comment is accurate, if a consumer updated the options, this would get passed through and updated because the useEffect handles if this prop changes:
https://github.com/cerner/terra-application/pull/2/files#diff-4de323c09ec8dd1f9bee39d9703e0872R37

File: src/disclosure-manager/DisclosureManagerContext.jsx
Comment (emilyrohrbough) [inline src/disclosure-manager/DisclosureManagerContext.jsx]: If a consumer provided their own `registerDismissCheck`, will it override the pending data checking?

File: src/disclosure-manager/DisclosureContainer.jsx
Comment (tbiethman) [reply src/disclosure-manager/DisclosureContainer.jsx]: Good call, that guy has been hiding out from an earlier implementation

File: src/modal-manager/ModalManager.jsx
Comment (tbiethman) [reply src/modal-manager/ModalManager.jsx]: It's accurate because DisclosureManager caches its disclosed components: https://github.com/cerner/terra-application/blob/updates-for-orion-application/src/disclosure-manager/DisclosureManager.jsx#L497

The DisclosureContainer is going to be created once with the current prop value and not updated again. So the comment is kind of accurate? Though an updated prop would apply to any subsequent disclosures...

File: src/disclosure-manager/DisclosureManagerContext.jsx
Comment (tbiethman) [reply src/disclosure-manager/DisclosureManagerContext.jsx]: It definitely would, for that disclosed component at least. The expectation is that components that override the default value would take responsibility over the pending actions.

I should probably call that out more explicitly in the documentation.

File: src/disclosure-manager/DisclosureContainer.jsx
Comment (tbiethman) [reply src/disclosure-manager/DisclosureContainer.jsx]: So just to clarify, the promptCheckpointRef is required for me to be able to call resolvePrompts at all. What *isn't* required is the registeredPromptsRef from the onPromptChange below, which I have removed.

File: src/disclosure-manager/DisclosureManagerContext.jsx
Comment (tbiethman) [reply src/disclosure-manager/DisclosureManagerContext.jsx]: Updated here: https://github.com/cerner/terra-application/pull/2/files#diff-feeb949f26ffe443e6fd52d1128878edR125

File: src/modal-manager/ModalManager.jsx
Comment (tbiethman) [reply src/modal-manager/ModalManager.jsx]: I could store the initial prop value in a ref and make things consistent throughout, but that seems a little heavy-handed. I will flesh out this comment a bit more to describe the potential inconsistency and best practices for use.

**Comment by emilyrohrbough** [comment]
@tbiethman Is the plan to deprecating the disclosure manager, slide-panel-manager, and SlidePanelManager?

File: src/disclosure-manager/DisclosureManagerDelegate.js
Comment (emilyrohrbough) [inline src/disclosure-manager/DisclosureManagerDelegate.js]: This is one-to-one with terra-disclosure-manager's DisclosureManagerDelegate. Do we need to reimplement or can we reuse the existing Delegate?

File: src/disclosure-manager/DisclosureManagerHeaderAdapter.jsx
Comment (emilyrohrbough) [inline src/disclosure-manager/DisclosureManagerHeaderAdapter.jsx]: This is one-to-one beside the `DisclosureManagerHeaderAdapterContext` import. Can we reuse the existing implementation?

File: src/disclosure-manager/index.js
Comment (emilyrohrbough) [inline src/disclosure-manager/index.js]: Should we provide this container to consumers or is this an internal implementation detail? Is the expected usage when a custom registerDismissCheck is used?

File: src/disclosure-manager/withDisclosureManager.jsx
Comment (emilyrohrbough) [inline src/disclosure-manager/withDisclosureManager.jsx]: I'm sure the answer is the same, but can we reuse the provided `withDisclosureManager` implementation? This is one-to-one

File: src/modal-manager/ModalManager.jsx
Comment (emilyrohrbough) [inline src/modal-manager/ModalManager.jsx]: Should we be maintaining the aria-label / does this affect screen readers? or was this removed due to lack of translations? https://github.com/cerner/terra-framework/blob/master/packages/terra-modal-manager/src/ModalManager.jsx#L85

File: src/modal-manager/ModalManager.module.scss
Comment (emilyrohrbough) [inline src/modal-manager/ModalManager.module.scss]: @tbiethman @ryanthemanuel  Is there someone working on filling in the new theme variables?

File: src/modal-manager/themes/orion-fusion-theme/orion-fusion-theme.scss
Comment (emilyrohrbough) [inline src/modal-manager/themes/orion-fusion-theme/orion-fusion-theme.scss]: Should these be applied as a global theme? Does this affect other themes by adding these as global? I would have thought these should be aggregated by the theme provider?

I'm looking at the port theme PR (https://github.com/cerner/terra-core/pull/2702) have will have the same questions there 😄

File: src/navigation-prompt/index.js
Comment (emilyrohrbough) [inline src/navigation-prompt/index.js:7]: missing prop descriptons

File: src/terra-dev-site/doc/FAQ.3.doc.mdx
Comment (emilyrohrbough) [inline src/terra-dev-site/doc/FAQ.3.doc.mdx]: nice faq

File: src/terra-dev-site/doc/how-to.2/consume-terra-application.0.doc.mdx
Comment (emilyrohrbough) [inline src/terra-dev-site/doc/how-to.2/consume-terra-application.0.doc.mdx]: > the recommended strategy for consuming `terra-application` is to define it as a `peerDependency` within your `package.json`.

It might be worth mentioning the application package / server level should be the only project that should define it as a hard dependency to reiterate: https://github.com/cerner/terra-application/pull/2/files#diff-dd3b6db2461ac3f1b54c8bd2cd30b6bfR7

File: src/disclosure-manager/DisclosureManager.jsx
Comment (emilyrohrbough) [inline src/disclosure-manager/DisclosureManager.jsx]: Would it be worth adding new disclosure manager props to the existing disclosure manger/slide-panel-manager & modal manger instead of reimplementing them?

```diff
// in terrra-disclosure-manager
....
- generateDisclosureComponentMappingForRender() {
+ generateDisclosureComponentMappingForRender(DisclosureContainer) {
    const { navigationPromptResolutionOptions } = this.props;

    const {
      disclosureComponentKeys,
      disclosureComponentData,
      disclosureComponentDelegates,
    } = this.state;

    return disclosureComponentKeys.reduce((accumulator, key, index) => {
      const componentData = disclosureComponentData[key];

+    const component = DisclosureContainer ?  <DisclosureContainer component={componentData.component} > : componentData.component;

      accumulator[key] = {
        component: (
          <DisclosureManagerHeaderAdapterContext.Provider value={componentData.headerAdapterContextValue} key={key}>
            <DisclosureManagerContext.Provider value={disclosureComponentDelegates[index]}>
-              {componentData.component}
+             {component}
            </DisclosureManagerContext.Provider>
          </DisclosureManagerHeaderAdapterContext.Provider>
        ),
        headerAdapterData: componentData.headerAdapterData,
      };

      return accumulator;
    }, {});
  }
...
  render() {
-    const { render, children } = this.props;
+    const { render, disclosureContainer, children } = this.props;

....
-    const disclosureComponentMappingForRender = this.generateDisclosureComponentMappingForRender();
+    const disclosureComponentMappingForRender = this.generateDisclosureComponentMappingForRender(disclosureContainer);
```

File: src/terra-dev-site/doc/how-to.2/show-loading-overlays.doc.mdx
Comment (emilyrohrbough) [inline src/terra-dev-site/doc/how-to.2/show-loading-overlays.doc.mdx]: Can consumers leverage the closed `ApplicationLoadingOverlayProvider` and still create an overlay relative to it's container? Or will the overlay be overall the content under the provider or will it default to full screen?

File: src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx
Comment (emilyrohrbough) [inline src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx]: Should we write these examples with the assumptions someone is already in an app that has already implemented ApplicationBase? Adding it here seems like a a copy-paste hazard where consumers will be implementing multiple layers of `ApplicationBase`

Comment (emilyrohrbough) [reply src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx]: Or should this how to for the base application level?

File: src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx
Comment (emilyrohrbough) [inline src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx]: of these navigationPromptResolutionOptions, are any of these required?

File: src/terra-dev-site/doc/how-to.2/use-translated-strings.doc.mdx
Comment (emilyrohrbough) [inline src/terra-dev-site/doc/how-to.2/use-translated-strings.doc.mdx]: nit, can we follow this syntax to mach what we typically do:
```
const ExampleComponent = ({ intl }) =>
...
export default injectIntl(ExampleComponent)
```

File: src/disclosure-manager/DisclosureManagerDelegate.js
Comment (tbiethman) [reply src/disclosure-manager/DisclosureManagerDelegate.js]: ~~Since we will be deprecating the existing packages, this is getting moved over wholesale.~~

Just correcting the record here. No, the packages won't be deprecated at this point in time. Changes have been made to the existing packages to facilitate the functionality we need here.

File: src/disclosure-manager/DisclosureManagerHeaderAdapter.jsx
Comment (tbiethman) [reply src/disclosure-manager/DisclosureManagerHeaderAdapter.jsx]: ~~Since we will be deprecating the existing packages, this is getting moved over wholesale.~~

Just correcting the record here. No, the packages won't be deprecated at this point in time. Changes have been made to the existing packages to facilitate the functionality we need here.

File: src/disclosure-manager/index.js
Comment (tbiethman) [reply src/disclosure-manager/index.js]: We could hide it from the index file, I don't expect people to use this directly.

File: src/disclosure-manager/withDisclosureManager.jsx
Comment (tbiethman) [reply src/disclosure-manager/withDisclosureManager.jsx]: ~~Since we will be deprecating the existing packages, this is getting moved over wholesale.~~

Just correcting the record here. No, the packages won't be deprecated at this point in time. Changes have been made to the existing packages to facilitate the functionality we need here.

File: src/modal-manager/ModalManager.jsx
Comment (tbiethman) [reply src/modal-manager/ModalManager.jsx]: I know the previous provided value "Modal" wasn't sufficient, and I don't know if even a translated string for "Modal" would be sufficient. It seems like this should be content-specific, but we have no way of setting this value from the disclosure API.

I could put the existing value back to keep things exactly the same and sync up with accessibility on a separate issue. But I don't think we want this to hold up the PR.

File: src/modal-manager/themes/orion-fusion-theme/orion-fusion-theme.scss
Comment (tbiethman) [reply src/modal-manager/themes/orion-fusion-theme/orion-fusion-theme.scss]: I was seeking consistency with that PRs implementation. I'll defer to decisions made there.

File: src/disclosure-manager/index.js
Comment (jmsv6d) [reply src/disclosure-manager/index.js]: Do we also want to prefix the filename with an underscore to indicate that it's private?

File: src/modal-manager/themes/orion-fusion-theme/orion-fusion-theme.scss
Comment (yuderekyu) [reply src/modal-manager/themes/orion-fusion-theme/orion-fusion-theme.scss]: We default to global in our webpack config. It doesn't hurt to be explicit. https://github.com/cerner/terra-toolkit/blob/master/config/webpack/webpack.config.js#L65

Yes, these files are aggregated to root or scoped files, which are then included as an entry point:
```
:root {
  @import ...modal-manager/themes...orion-fusion-theme.scss
}
``` 
```
.orion-fusion-theme {
  @import ...modal-manager/themes...orion-fusion-theme.scss
}
``` 
Scoping these globally avoids a second scope layer. This should not be an issue because theme files are not imported in the component. Someone could manually import these, but that's on them.

**Review by emilyrohrbough — CHANGES_REQUESTED** [review]
blocked by https://github.com/cerner/terra-framework/pull/973

File: src/terra-dev-site/doc/how-to.2/change-application-theme.doc.mdx
Comment (yuderekyu) [inline src/terra-dev-site/doc/how-to.2/change-application-theme.doc.mdx]: not necessary for this pr, but we should link to our theme aggregation docs

**Review by emilyrohrbough — APPROVED** [review]

File: src/modal-manager/ModalManager.jsx
Comment (emilyrohrbough) [inline src/modal-manager/ModalManager.jsx:29]: This is clean 😄

**Comment by mjhenkes** [comment]
You're going to be really annoyed at me, but could you put all your doc in a 'terra-application' folder? 😬

File: src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx
Comment (mjhenkes) [reply src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx]: I like how the example separates the application setup from the component doing the disclosure and I think all it needs is a little more description about 'this is how the app is setup and you may not need to render a modal manager if you're a child of a component that already rendered it'...

File: src/terra-dev-site/doc/reference.1/components/NavigationPromptCheckpoint.doc.mdx
Comment (mjhenkes) [inline src/terra-dev-site/doc/reference.1/components/NavigationPromptCheckpoint.doc.mdx]: any of these props required?

File: src/terra-dev-site/doc/how-to.2/change-application-theme.doc.mdx
Comment (mjhenkes) [inline src/terra-dev-site/doc/how-to.2/change-application-theme.doc.mdx]: Does this page need more discussion about how to set a default theme and IE10 Theme change limitations?

File: src/terra-dev-site/doc/reference.1/components/ApplicationBase.doc.mdx
Comment (mjhenkes) [inline src/terra-dev-site/doc/reference.1/components/ApplicationBase.doc.mdx]:
```suggestion
> Note: ApplicationBase satisfies the base-level requirements for Terra component usage. Application base encapsulates the `terra-base` component making it unnecessary to render a separate instance.
```

File: src/terra-dev-site/doc/reference.1/components/NavigationPromptCheckpoint.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/reference.1/components/NavigationPromptCheckpoint.doc.mdx]: Not technically required. If nothing is provided, you're going to get a blank dialog with two blank buttons with no emphasis. I can't imagine that's ideal, but I'm not sure we should enforce any one thing to be required here.

File: src/modal-manager/ModalManager.jsx
Comment (tbiethman) [reply src/modal-manager/ModalManager.jsx]: This is has been resolved by later changes 👍 The updated prop value should propagate through appropriately.

File: src/disclosure-manager/index.js
Comment (tbiethman) [reply src/disclosure-manager/index.js]: I can make that happen

File: src/modal-manager/ModalManager.module.scss
Comment (tbiethman) [reply src/modal-manager/ModalManager.module.scss]: This has been removed.

**Review by StephenEsser — APPROVED** [review]

File: src/modal-manager/ModalManager.jsx
Comment (tbiethman) [reply src/modal-manager/ModalManager.jsx]: Since we're using the existing ModalManager package now, this was not changed.

File: src/terra-dev-site/doc/how-to.2/show-loading-overlays.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/how-to.2/show-loading-overlays.doc.mdx]: The location of the overlay entirely dependent upon the location of the ApplicationLoadingOverlayProvider in the component hierarchy. If a provider is presenting an overlay, it will overlay all content rendered within the provider.

File: src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx]: I'll elaborate

File: src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/how-to.2/show-modal-content.doc.mdx]: Not technically, no. I don't think we're able to establish hard requirements around this within terra-application at this point in time. I'd like to be able to at some point.

File: src/terra-dev-site/doc/how-to.2/use-translated-strings.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/how-to.2/use-translated-strings.doc.mdx]: Honestly, I'd rather promote this pattern.

File: src/terra-dev-site/doc/reference.1/components/ApplicationBase.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/reference.1/components/ApplicationBase.doc.mdx]: It works like a charm

File: src/terra-dev-site/doc/how-to.2/change-application-theme.doc.mdx
Comment (tbiethman) [reply src/terra-dev-site/doc/how-to.2/change-application-theme.doc.mdx]: Probably, but we're going to have to follow up on that later.

**Pull request merged**
